src/plugins/dashboard/server/api/register_routes.ts

```diff
@@ -0,0 +1,319 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the "Elastic License
+ * 2.0", the "GNU Affero General Public License v3.0 only", and the "Server Side
+ * Public License v 1"; you may not use this file except in compliance with, at
+ * your election, the "Elastic License 2.0", the "GNU Affero General Public
+ * License v3.0 only", or the "Server Side Public License, v 1".
+ */
+
+import { schema } from '@kbn/config-schema';
+import type { ContentManagementServerSetup } from '@kbn/content-management-plugin/server';
+import type { HttpServiceSetup } from '@kbn/core/server';
+import type { UsageCounter } from '@kbn/usage-collection-plugin/server';
+import type { Logger } from '@kbn/logging';
+
+import { CONTENT_ID } from '../../common/content_management';
+import {
+  PUBLIC_API_PATH,
+  PUBLIC_API_VERSION,
+  PUBLIC_API_CONTENT_MANAGEMENT_VERSION,
+} from './constants';
+import {
+  dashboardAttributesSchema,
+  dashboardGetResultSchema,
+  dashboardCreateResultSchema,
+  dashboardSearchResultsSchema,
+  referenceSchema,
+} from '../content_management/v3';
+
+interface RegisterAPIRoutesArgs {
+  http: HttpServiceSetup;
+  contentManagement: ContentManagementServerSetup;
+  restCounter?: UsageCounter;
+  logger: Logger;
+}
+
+function recursiveSortObjectByKeys(obj: unknown): unknown {
+  if (typeof obj !== 'object' || obj === null) return obj;
+  if (Array.isArray(obj)) {
+    return obj.map(recursiveSortObjectByKeys);
+  }
+  return Object.keys(obj)
+    .sort()
+    .reduce((acc, key) => {
+      acc[key] = recursiveSortObjectByKeys(obj[key]);
+      return acc;
+    }, {} as Record<string, unknown>);
+}
+
+export function registerAPIRoutes({
+  http,
+  contentManagement,
+  restCounter,
+  logger,
+}: RegisterAPIRoutesArgs) {
+  const { versioned: versionedRouter } = http.createRouter();
+
+  // TODO add usage collection
+
+  // TODO add authorization checks
+
+  // Create API route
+  const createRoute = versionedRouter.post({
+    path: `${PUBLIC_API_PATH}/{id?}`,
+    access: 'public',
+    description: `Create a new dashboard.`,
```

Comment:
Not a blocker: for these routes I'd recommend adding an OAS tag for these endpoints

```suggestion
    description: `Create a new dashboard.`,
    options: {
      tags: ['oas-tag:Dashboards'],
    },
```

That way they can appear in their own group in our [hosted docs](https://www.elastic.co/docs/api/doc/kibana/v8).